pkg/api/dirs.go

```diff
@@ -0,0 +1,155 @@
+	if err != nil {
+		s.Logger.Errorf("dir upload, store dir: %v", err)
+		jsonhttp.InternalServerError(w, "could not store dir")
+		return
+	}
+
+	jsonhttp.OK(w, DirUploadResponse{
+		Reference: reference,
+	})
+}
+
+// getDirHTTPInfo extracts data for a directory to be uploaded from an HTTP request
+func getDirHTTPInfo(r *http.Request) (*dirUploadInfo, error) {
+	if r.Body == http.NoBody {
+		return &dirUploadInfo{}, errors.New("request has no body")
+	}
+	toEncrypt := strings.ToLower(r.Header.Get(encryptHeader)) == "true"
+	return &dirUploadInfo{
+		reader:    r.Body,
+		toEncrypt: toEncrypt,
+	}, nil
+}
+
+// storeDir stores all files recursively contained in the directory given as a tar
+// it returns the hash for the uploaded manifest corresponding to the uploaded dir
+func storeDir(ctx context.Context, dirInfo *dirUploadInfo, s storage.Storer, logger logging.Logger) (swarm.Address, error) {
+	dirManifest := jsonmanifest.NewManifest()
+
+	// set up HTTP body reader
+	tarReader := tar.NewReader(dirInfo.reader)
+	defer dirInfo.reader.Close()
+
+	// iterate through the files in the supplied tar
+	for {
+		fileHeader, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return swarm.ZeroAddress, fmt.Errorf("read tar stream error: %v", err)
+		}
+
+		filePath := fileHeader.Name
+
+		// only store regular files
+		if !fileHeader.FileInfo().Mode().IsRegular() {
+			logger.Warningf("skipping file upload for %s as it is not a regular file", filePath)
+			continue
+		}
+
+		fileName := fileHeader.FileInfo().Name()
+		contentType := mime.TypeByExtension(filepath.Ext(fileHeader.Name))
+
+		// upload file
+		fileInfo := &FileUploadInfo{
+			name:        fileName,
+			size:        fileHeader.FileInfo().Size(),
+			contentType: contentType,
+			toEncrypt:   dirInfo.toEncrypt,
+			reader:      tarReader,
+		}
+		fileReference, err := storeFile(ctx, fileInfo, s)
+		if err != nil {
+			return swarm.ZeroAddress, fmt.Errorf("store dir file error: %v", err)
+		}
+		logger.Tracef("uploaded dir file %v with reference %v", filePath, fileReference)
+
+		// create manifest entry for uploaded file
+		headers := http.Header{}
+		headers.Set("Content-Type", contentType)
+		fileEntry := &jsonmanifest.JSONEntry{
+			Reference: fileReference,
+			Name:      fileName,
+			Headers:   headers,
+		}
+
+		// add entry to dir manifest
+		dirManifest.Add(filePath, fileEntry)
+	}
+
+	// check if files were uploaded through the manifest
+	if len(dirManifest.Entries) == 0 {
+		return swarm.ZeroAddress, fmt.Errorf("no files in tar")
+	}
+
+	// upload manifest
+	// first, serialize into byte array
+	b, err := dirManifest.Serialize()
+	if err != nil {
+		return swarm.ZeroAddress, fmt.Errorf("manifest serialize error: %v", err)
+	}
+
+	// set up reader for manifest file upload
+	r := bytes.NewReader(b)
+
+	// then, upload manifest
+	manifestFileInfo := &FileUploadInfo{
+		name:        "manifest.json",
```

Comment:
There is no special reason to have name for this file.
It basically has no difference. It just would not be correct for someone to assume that manifest file needs to be called like this. Especially since we are to move away from JSON structure.

Reply:
fixed, thanks